text/0009-BundleTracker-integration.md

```diff
@@ -0,0 +1,588 @@
+
+The `BundleTrackerCustomizer` methods offer the ability to create and manage custom objects to be tracked (the pointer returned from `AddingBundle(const Bundle&, const BundleEvent&)` will be passed into the other methods when called on the same `Bundle`). Additionally, if a `NULL` is returned, the object will not be tracked by the `BundleTracker`.
+
+The three callbacks are called under the following conditions:
+
+`AddingBundle`:
+
+- A `Bundle` enters a state covered by the `stateMask`, and is not currently being tracked.
+
+`RemovedBundle`:
+
+- A `Bundle` that is currently being tracked enters a state not covered by the `stateMask`.
+- A `Bundle` is removed manually using `Remove()`
+- A `Bundle` is currently being tracked when the `Close()` method is called on the `BundleTracker` (or when the `BundleTracker` goes out of scope).
+
+`ModifiedBundle`:
+
+- A `Bundle` enters a state covered by the `stateMask`, and is currently being tracked.
+
+#### Extending a BundleTracker
+
+The [OSGi specification](http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html#d0e52020) permits extension of the `BundleTracker` by using wrapper objects. This behavior is mapped to C++ using a template parameter for the `BundleTracker`. By specifying a wrapper object of type `T` to the `BundleTracker`, the user can make the tracker hold on to custom objects through the tracking callbacks that contain extra data or functionality.
+
+## Detailed Design
+****
+### Overview
+
+The `BundleTracker` is implemented as part of of the core framework for CppMicroservices, leveraging existing constructs to listen for fired `BundleEvent` objects and maintaining the map of tracked objects.
+
+The `BundleTracker` uses a pointer-to-implementation pattern to hide data members and implementation details from the user. This information is contained in `BundleTrackerPrivate`, which has a one-to-one relationship with the `BundleTracker` through a `std::unique_ptr`. This construct in turn owns an instance of the `TrackedBundle`, which subclasses `BundleAbstractTracked` and handles tracking operations while the `BundleTracker` is open.
+
+The customization scheme comes from a back pointer from the `TrackedBundle` class. When a `BundleTrackerCustomizer` is provided by the user, the pointer points to it. When `BundleTrackerCustomizer` is NULL when constructed, it points to the `BundleTracker` itself to use the default or overridden methods.
+
+It is worth noting that the API for `BundleTracker` is different from the current implementation of the `ServiceTracker`, which still uses a raw pointer for the `ServiceTracker` constructor's `ServiceTrackerCustomizer` argument. In the future, the `ServiceTracker` API should be adapted to function like the new `BundleTracker` API to make the lifetime of the `ServiceTrackerCustomizer` more explicit.
```

Comment:
2 comments:
- I suggest putting the reason for the difference earlier in this paragraph.
- The reason to use `std::shared_ptr` over a raw pointer is to guarantee the lifetime of the `ServiceTrackerCustomizer` object is scoped to the lifetime of the `ServiceTracker` object.  "making the lifetime explicit" is a little vague.